File: ESTRUCURAS_DE_DATOS/LISTAS_DOBLES_ENLAZADAS/PYTHON/listasDobles.py
# Doubly linked list - versión corregida y sencilla

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def append(self, value):
        """Agregar al final"""
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        self.length += 1

    def prepend(self, value):
        """Agregar al inicio"""
        new_node = Node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        self.length += 1

    def traverse_to_index(self, index):
        """Devuelve el nodo en la posición index (0-based). Lanza IndexError si fuera de rango."""
        if index < 0 or index >= self.length:
            raise IndexError("Index out of range")
        current = self.head
        i = 0
        while i < index:
            current = current.next
            i += 1
        return current

    def insert(self, index, value):
        """Inserta en la posición index (0 = inicio). Si index >= length -> append."""
        if index <= 0:
            self.prepend(value)
            return
        if index >= self.length:
            self.append(value)
            return
        # insertar en medio
        new_node = Node(value)
        leader = self.traverse_to_index(index - 1)
        follower = leader.next
        leader.next = new_node
        new_node.prev = leader
        new_node.next = follower
        if follower:
            follower.prev = new_node
        self.length += 1

    def remove(self, index):
        """Elimina el nodo en la posición index (0-based). Lanza IndexError si fuera de rango."""
        if self.length == 0:
            raise IndexError("Remove from empty list")
        if index < 0 or index >= self.length:
            raise IndexError("Index out of range")

        if index == 0:
            # eliminar cabeza
            removed = self.head
            self.head = self.head.next
            if self.head:
                self.head.prev = None
            else:
                # lista quedó vacía
                self.tail = None
            self.length -= 1
            return removed.value

        if index == self.length - 1:
            # eliminar cola
            removed = self.tail
            self.tail = self.tail.prev
            if self.tail:
                self.tail.next = None
            else:
                self.head = None
            self.length -= 1
            return removed.value

        leader = self.traverse_to_index(index - 1)
        unwanted = leader.next
        follower = unwanted.next
        leader.next = follower
        if follower:
            follower.prev = leader
        self.length -= 1
        return unwanted.value

    def to_list_forward(self):
        result = []
        current = self.head
        while current:
            result.append(current.value)
            current = current.next
        return result

    def to_list_backward(self):
        result = []
        current = self.tail
        while current:
            result.append(current.value)
            current = current.prev
        return result

    def print_list(self):
        vals = self.to_list_forward()
        if not vals:
            print("(empty)")
        else:
            print(" <-> ".join(str(v) for v in vals))


# ---------- Código de prueba ----------
if __name__ == "__main__":
    my_list = DoublyLinkedList()
    my_list.append(10)
    my_list.append(20)
    my_list.append(5)
    my_list.append(6)

    print("Inicial:")
    my_list.print_list()              # 10 <-> 20 <-> 5 <-> 6

    print("\nInsertar 15 en pos 2:")
    my_list.insert(2, 15)
    my_list.print_list()              # 10 <-> 20 <-> 15 <-> 5 <-> 6

    print("\nPrepend 3:")
    my_list.prepend(3)
    my_list.print_list()              # 3 <-> 10 <-> 20 <-> 15 <-> 5 <-> 6

    print("\nInsertar 25 en pos 3:")
    my_list.insert(3, 25)
    my_list.print_list()              # 3 <-> 10 <-> 20 <-> 25 <-> 15 <-> 5 <-> 6

    print("\nEliminar índice 7 (último):")
    try:
        removed = my_list.remove(my_list.length - 1)
        print("Eliminado:", removed)
    except IndexError as e:
        print("Error:", e)
    my_list.print_list()

File: ESTRUCURAS_DE_DATOS/LISTAS_DOBLES_ENLAZADAS/typeScript/listasDobles.ts
// ------------------ Nodo ------------------
class Nodo<T> {
    value: T;
    next: Nodo<T> | null;
    prev: Nodo<T> | null;

    constructor(value: T) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

// ------------------ Lista Doblemente Enlazada ------------------
class DoublyLinkedList<T> {
    head: Nodo<T> | null;
    tail: Nodo<T> | null;
    length: number;

    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    // Agregar al final
    append(value: T): void {
        const newNode = new Nodo(value);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.prev = this.tail;
            if (this.tail) this.tail.next = newNode;
            this.tail = newNode;
        }
        this.length++;
    }

    // Agregar al inicio
    prepend(value: T): void {
        const newNode = new Nodo(value);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            newNode.next = this.head;
            this.head.prev = newNode;
            this.head = newNode;
        }
        this.length++;
    }

    // Obtener nodo en una posición específica
    traverseToIndex(index: number): Nodo<T> {
        if (index < 0 || index >= this.length) {
            throw new Error("Index out of range");
        }

        let current = this.head;
        let i = 0;
        while (current && i < index) {
            current = current.next;
            i++;
        }

        if (!current) throw new Error("Index not found");
        return current;
    }

    // Insertar en una posición
    insert(index: number, value: T): void {
        if (index <= 0) {
            this.prepend(value);
            return;
        }
        if (index >= this.length) {
            this.append(value);
            return;
        }

        const newNode = new Nodo(value);
        const leader = this.traverseToIndex(index - 1);
        const follower = leader.next;

        leader.next = newNode;
        newNode.prev = leader;
        newNode.next = follower;

        if (follower) follower.prev = newNode;

        this.length++;
    }

    // Eliminar un nodo en un índice
    remove(index: number): T {
        if (this.length === 0) {
            throw new Error("Remove from empty list");
        }
        if (index < 0 || index >= this.length) {
            throw new Error("Index out of range");
        }

        let removed: Nodo<T> | null = null;

        if (index === 0) {
            removed = this.head;
            this.head = this.head ? this.head.next : null;
            if (this.head) this.head.prev = null;
            else this.tail = null;
        } else if (index === this.length - 1) {
            removed = this.tail;
            this.tail = this.tail ? this.tail.prev : null;
            if (this.tail) this.tail.next = null;
            else this.head = null;
        } else {
            const leader = this.traverseToIndex(index - 1);
            removed = leader.next;
            const follower = removed ? removed.next : null;
            leader.next = follower;
            if (follower) follower.prev = leader;
        }

        this.length--;
        return removed ? removed.value : (undefined as any);
    }

    // Convertir a array (de cabeza a cola)
    toListForward(): T[] {
        const result: T[] = [];
        let current = this.head;
        while (current) {
            result.push(current.value);
            current = current.next;
        }
        return result;
    }

    // Convertir a array (de cola a cabeza)
    toListBackward(): T[] {
        const result: T[] = [];
        let current = this.tail;
        while (current) {
            result.push(current.value);
            current = current.prev;
        }
        return result;
    }

    // Imprimir la lista
    printList(): void {
        const vals = this.toListForward();
        if (vals.length === 0) console.log("(empty)");
        else console.log(vals.join(" <-> "));
    }
}

// ------------------ Código de prueba ------------------
const myList = new DoublyLinkedList<number>();
myList.append(10);
myList.append(20);
myList.append(5);
myList.append(6);

console.log("Inicial:");
myList.printList(); // 10 <-> 20 <-> 5 <-> 6

console.log("\nInsertar 15 en pos 2:");
myList.insert(2, 15);
myList.printList(); // 10 <-> 20 <-> 15 <-> 5 <-> 6

console.log("\nPrepend 3:");
myList.prepend(3);
myList.printList(); // 3 <-> 10 <-> 20 <-> 15 <-> 5 <-> 6

console.log("\nInsertar 25 en pos 3:");
myList.insert(3, 25);
myList.printList(); // 3 <-> 10 <-> 20 <-> 25 <-> 15 <-> 5 <-> 6

console.log("\nEliminar último elemento:");
const removed = myList.remove(myList.length - 1);
console.log("Eliminado:", removed);
myList.printList();
